azurerm/internal/services/apimanagement/api_management_gateway_api_resource.go

```diff
@@ -0,0 +1,162 @@
+	if err != nil {
+		if !utils.ResponseWasStatusCode(exists, http.StatusNoContent) {
+			if !utils.ResponseWasNotFound(exists) {
+				return fmt.Errorf("checking for presence of existing API %q / Gateway %q (API Management Service %q / Resource Group %q): %+v", apiID.Name, gatewayID, gatewayID.ServiceName, gatewayID.ResourceGroup, err)
+			}
+		}
+	}
+
+	if !utils.ResponseWasNotFound(exists) {
+		// TODO: can we pull this from somewhere?
+		subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+		resourceId := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ApiManagement/service/%s/gateways/%s/apis/%s", subscriptionId, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID, apiID.Name)
+		return tf.ImportAsExistsError("azurerm_api_management_gateway_api", resourceId)
+	}
+
+	params := &apimanagement.AssociationContract{}
+	resp, err := client.CreateOrUpdate(ctx, gatewayID.ResourceGroup, gatewayID.ServiceName, gatewayID.Name, apiID.Name, params)
+	if err != nil {
+		return fmt.Errorf("adding API %q to Gateway %q (API Management Service %q / Resource Group %q): %+v", apiID.Name, gatewayID.Name, gatewayID.ServiceName, gatewayID.ResourceGroup, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceApiManagementGatewayApiRead(d, meta)
+}
+
+func resourceApiManagementGatewayApiRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).ApiManagement.GatewayApisClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.GatewayApiID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	serviceName := id.ServiceName
+	gatewayID := id.GatewayName
+	apiName := id.ApiName
+
+	apiId := parse.NewApiID(id.SubscriptionId, id.ResourceGroup, id.ServiceName, id.ApiName)
+
+	resp, err := client.GetEntityTag(ctx, resourceGroup, serviceName, gatewayID, apiName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp) {
+			log.Printf("[DEBUG] gateway %q (API Management Service %q / Resource Group %q) does not exist - removing from state!", apiName, serviceName, resourceGroup)
+			d.SetId("")
+			return nil
+		}
+
+		if utils.ResponseWasStatusCode(resp, http.StatusNoContent) {
+			log.Printf("[DEBUG] gateway %q (API Management Service %q / Resource Group %q) returning a no content status - bypassing and moving on!", apiName, serviceName, resourceGroup)
+		} else {
+			return fmt.Errorf("retrieving gateway %q (API Management Service %q / Resource Group %q): %+v", apiName, serviceName, resourceGroup, err)
+		}
+	}
+
+	if utils.ResponseWasNotFound(resp) {
+		log.Printf("[DEBUG] API %q was not found in Gateway  %q (API Management Service %q / Resource Group %q) was not found - removing from state!", apiName, gatewayID, serviceName, resourceGroup)
+		d.SetId("")
+		return nil
+	}
+
+	gateway := parse.NewGatewayID(id.SubscriptionId, resourceGroup, serviceName, gatewayID)
```

Comment:
As above, we can simplify this by referencing ID fields directly and making use of ID struct String methods.

I'm assuming the "gatewayID" and "apiid" params to client.GetEntityTag are actually the gateway name and API name respectively?

```suggestion

	apiId := parse.NewApiID(id.SubscriptionId, id.ResourceGroup, id.ServiceName, id.ApiName)

	resp, err := client.GetEntityTag(ctx, id.ResourceGroup, id.ServiceName, id.GatewayName, id.ApiName)
	if err != nil {
		if utils.ResponseWasNotFound(resp) {
			log.Printf("[DEBUG] %s does not exist - removing from state!", id)
			d.SetId("")
			return nil
		}

		if utils.ResponseWasStatusCode(resp, http.StatusNoContent) {
			log.Printf("[DEBUG] %s returned with No Content status - bypassing and moving on!", id)
		} else {
			return fmt.Errorf("retrieving %s: %+v", id, err)
		}
	}

	if utils.ResponseWasNotFound(resp) {
		log.Printf("[DEBUG] %s was not found - removing from state!", id)
		d.SetId("")
		return nil
	}

	gateway := parse.NewGatewayID(id.SubscriptionId, id.ResourceGroup, id.ServiceName, id.GatewayName)
```

Reply:
updated